modules/_labs/color-picker/react/lib/ColorPicker.tsx

```diff
@@ -0,0 +1,192 @@
+  colors.fruitPunch400,
+  colors.rootBeer400,
+  colors.toastedMarshmallow400,
+  colors.licorice400,
+  colors.cappuccino400,
+  colors.blackPepper400,
+];
+
+const Container = styled('div')({
+  width: 224,
+  padding: spacing.s,
+  display: 'flex',
+  flexDirection: 'column',
+  alignItems: 'center',
+  background: colors.frenchVanilla100,
+  border: `1px solid ${colors.soap400}`,
+  borderRadius: borderRadius.m,
+  ...depth['3'],
+});
+
+const ColorInputWrapper = styled('div')({
+  width: '100%',
+  marginTop: spacing.s,
+  display: 'flex',
+  flexDirection: 'row',
+  justifyContent: 'space-between',
+});
+
+const ColorInputAndLabel = styled(FormField)({
+  display: 'flex',
+  flexDirection: 'column',
+  margin: 0,
+});
+
+const CheckButton = styled(IconButton)({
+  alignSelf: 'flex-end',
+});
+
+const HexColorInput = styled(ColorInput)({
+  width: '168px',
+});
+
+const isCustomColor = (colors: string[], hexCode?: string) => {
+  if (!hexCode) {
+    return false;
+  }
+
+  const lowercaseHex = hexCode.toLowerCase();
+  return !colors.includes(lowercaseHex);
+};
+
+export const ColorPicker = ({
+  colorSet = defaultColors,
+  customHexInputLabel = 'Custom Hex Color',
+  onColorChange,
+  onColorReset,
+  onSubmitClick,
+  resetLabel = 'Reset',
+  resetColor,
+  value = '',
+  showCustomHexInput = false,
+  ...elemProps
+}: ColorPickerProps) => {
+  const [validHexValue, setValidHexValue] = React.useState('');
+  const [customHexValue, setCustomHexValue] = React.useState(
+    isCustomColor(colorSet, value) ? value : ''
+  );
+
+  const onCustomHexChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setCustomHexValue(event.target.value);
+    setValidHexValue('');
+  };
+
+  const onValidHexChange = (validHexValue: string) => {
+    setValidHexValue(validHexValue);
+  };
+
+  const onCustomInputKeyDown = (event: React.KeyboardEvent) => {
+    if (event.key === 'Enter' && validHexValue) {
+      onColorChange(validHexValue);
+    }
+  };
+
+  const placeholder = customHexValue === '' ? undefined : customHexValue.toUpperCase();
+
+  return (
+    <Container {...elemProps}>
+      {onColorReset && resetColor && (
+        <ResetButton onClick={onColorReset} resetColor={resetColor} label={resetLabel} />
+      )}
+      <SwatchBook colors={colorSet} onSelect={onColorChange} value={value} />
+      {showCustomHexInput && (
+        <ColorInputWrapper>
+          <ColorInputAndLabel label={customHexInputLabel}>
+            <HexColorInput
+              onChange={onCustomHexChange}
+              onKeyDown={onCustomInputKeyDown}
+              onValidColorChange={onValidHexChange}
+              value={customHexValue}
+              placeholder={placeholder}
+              showCheck={!!placeholder}
+            />
+          </ColorInputAndLabel>
+          <CheckButton
+            aria-label={customHexInputLabel}
+            variant={IconButton.Variant.CircleFilled}
+            icon={checkIcon}
+            onClick={onSubmitClick}
```

Comment:
Shouldn't this basically just call `onColorChange`? I don't really see the value in this callback right now. This is the reason there are bugs in the stories (regular colors and custom colors being selected at the same time, etc.)

Reply:
ah, yeah. the only reason it was separate is because the designs say that button should close the popup, so that's how I chose to accomplish that. feel free to change it!

Reply:
Ah makes sense. In that case, I think we need to call both `onColorChange` and `onSubmitClick` as long as the custom color is valid.